ci/scripts/check-bazel-tags.sh

```diff
@@ -6,14 +6,103 @@
+check_empty () {
+if [[ ${1} ]]; then
+  echo "Error:"
+  echo "$1"|sed 's/^/    /';
+  echo "$2"
   exit 1
 fi
+}
+
+# This check ensures OpenTitan software can be built with a wildcard without
+# waiting for Verilator using --build_tag_filters=-verilator
+untagged=$(./bazelisk.sh query \
+  "rdeps(
+      //...,
+      //hw:verilator
+  )
+  except
+  attr(
+      tags,
+      '$(exact_regex "(verilator|manual)")',
+      //...
+  )" \
+  --output=label_kind)
+check_empty "${untagged}" \
+"Target(s) above depend(s) on //hw:verilator, please tag it with verilator or manual."
+
+# This check ensures Opentitan software can be built with wildcards in
+# environments that don't have vivado or vivado tools installed by using
+# --build_tag_filters=-vivado.
+untagged=$(./bazelisk.sh query \
+  "rdeps(
+      //...,
+      kind(
+          'bitstream_splice',
+          //...
+      )
+  )
+  except
+  attr(
+      tags,
+      '$(exact_regex "(cw310_rom|cw310_test_rom|vivado|manual)")',
+      //...
+  )" \
+  --output=label_kind)
+check_empty "${untagged}" \
+"Target(s) above depend(s) on a bitstream_splice and isn't available in a cache.
+Please tag it with vivado or manual."
+
+# This check ensures cw310 users may group tests that depend on the cached
+# cw310_test_rom bitstream.
+mistagged=$(./bazelisk.sh query \
+    "tests(`# Only output tests`
+        rdeps(
+            attr(`# Anything tagged cw310_test_rom`
+                tags,
+                '$(exact_regex cw310_test_rom)',
+                tests(//...)
+            ),
+            kind(`# That depends on a bitstream splice`
+                'bitstream_splice',
+                //...
+            )
+            except`# Other than those used to build the test_ROM bitrearm`
+            deps(
+                //hw/bitstream:test_rom
+            )
+        )
+    )" \
+    --output=label_kind)
+check_empty "${mistagged}" \
+"Target(s) above depend(s) on a bitstream_splice other than those used to generate the test_rom.
+Please tag as cw310_*_variant if you intended to use another bitstream."
+
+# This check ensures cw310 users may group tests that depend on the cached
+# cw310_rom bitstream.
+mistagged=$(./bazelisk.sh query \
+    "tests(`# Only output tests`
+        rdeps(
+            attr(`# Anything tagged cw310_rom`
+                tags,
+                '$(exact_regex cw310_rom)',
+                //...
+            ),
+            kind(`# That depends on a bitstream splice`
+                'bitstream_splice',
+                //...
+            )
+            except`# Other than those used to build the ROM bitstream`
+            deps(//hw/bitstream:rom)
+        )
+    )" \
+    --output=label_kind)
+check_empty "${mistagged}" \
+"Above target(s) depend(s) on a bitstream_splice other than those used to generate the rom.
+Please tag as cw310_*_variant if you intend to use another bitstream."
```

Comment:
Same as the comment above re: a more explicit error message:

```suggestion
"The target(s) above are tagged with cw310_rom but depend(s) on a bitstream_splice rule other than those used to generate the rom.
Please tag with cw310_*_variant instead if you intended to use another bitstream."
```

Reply:
I can clean that up a little. The tag is also usually implicitly set based on the target and I should clarify that.